### PR TITLE
Update DefaultNoBound derive macro

### PR DESCRIPTION
**Pull request by benluelo**
Updates the `DefaultNoBound` derive macro.
- Fixes a bug when deriving for enums with no generics
  - This may seem like an odd use case, but I hit this when deriving `DefaultNoBound` in generated code
- Changes the behaviour on enums for selecting the default variant from silently choosing the first variant to requiring a `#[default]` attribute. This aligns with std: https://github.com/rust-lang/rust/issues/87517
- Small improvements to diagnostics:
  - Fixed erroneous mention of `CloneNoBound` when attempting to derive on a `union` (and add UI test)
  - Added UI tests for the new `#[default]` attribute
  - Added compile pass tests for deriving `DefaultNoBound` on a `struct` and `enum` with no generics
- Cleaned up the code a bit, importing items instead of using fully qualified paths and renaming single letter variables

Polkadot address: 12xTJPv5HhLeqApdwpP9RMiLhvNpfF2HgEW6gDtMbsCqair8

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2056575

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2056577

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2056582

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2056576

File: frame/support/procedural/src/default_no_bound.rs [hunk 1]
```diff
@@ -25,72 +26,122 @@ pub fn derive_default_no_bound(input: proc_macro::TokenStream) -> proc_macro::To
...
+			let tagged_variants = enum_
+				.variants
+				.into_iter()
+				.flat_map(|v| {
+					v.attrs
+						.clone()
+						.into_iter()
+						.filter(|a| a.path.is_ident("default"))
+						.map(move |_| v.clone())
+				})
```
Comment (bkchr) [inline frame/support/procedural/src/default_no_bound.rs:70]:
```suggestion
				.filter(|v| {
					v.attrs
						.iter()
						.any(|a| a.path.is_ident("default"))
				})
```

File: frame/support/procedural/src/default_no_bound.rs [hunk 2]
```diff
@@ -25,72 +26,122 @@ pub fn derive_default_no_bound(input: proc_macro::TokenStream) -> proc_macro::To
...
-						let fields = unnamed
-							.unnamed
-							.iter()
-							.enumerate()
-							.map(|(i, _)| syn::Index::from(i))
-							.map(|i| {
-								quote::quote_spanned!(i.span() =>
+		Data::Enum(enum_) => {
+			if enum_.variants.is_empty() {
+				return quote! { Self }.into();
```
Comment (bkchr) [inline frame/support/procedural/src/default_no_bound.rs:58]: The return is wrong here. It would make the implementation be `Self`.

File: frame/support/procedural/src/default_no_bound.rs [hunk 3]
```diff
@@ -25,72 +26,122 @@ pub fn derive_default_no_bound(input: proc_macro::TokenStream) -> proc_macro::To
...
+				[] => {
+					return syn::Error::new(name.clone().span(), "no #[default] attribute found")
+						.into_compile_error()
+						.into()
+				},
+				[default_variant] => {
+					let variant_ident = &default_variant.ident;
+
+					let fully_qualified_variant_path =
+						quote! { <#name #ty_generics>::#variant_ident };
```
Comment (bkchr) [inline frame/support/procedural/src/default_no_bound.rs:83]:
```suggestion
						quote! { Self::#variant_ident };
```

File: frame/support/procedural/src/default_no_bound.rs [hunk 4]
```diff
@@ -25,72 +26,122 @@ pub fn derive_default_no_bound(input: proc_macro::TokenStream) -> proc_macro::To
...
+
+					match &default_variant.fields {
+						Fields::Named(named) => {
+							let fields =
+								named.named.iter().map(|field| &field.ident).map(|ident| {
+									quote_spanned! {ident.span()=>
```
Comment (bkchr) [inline frame/support/procedural/src/default_no_bound.rs:89]:
```suggestion
									quote_spanned! { ident.span()=>
```

File: frame/support/procedural/src/default_no_bound.rs [hunk 5]
```diff
@@ -25,72 +26,122 @@ pub fn derive_default_no_bound(input: proc_macro::TokenStream) -> proc_macro::To
...
+										#ident: core::default::Default::default()
+									}
+								});
+
+							quote! { #fully_qualified_variant_path { #( #fields, )* } }
+						},
+						Fields::Unnamed(unnamed) => {
+							let fields = unnamed.unnamed.iter().map(|field| {
+								quote_spanned! {field.span()=>
```
Comment (bkchr) [inline frame/support/procedural/src/default_no_bound.rs:98]:
```suggestion
								quote_spanned! { field.span()=>
```

File: frame/support/procedural/src/default_no_bound.rs (hunk 4 above)
Comment (benluelo) [reply frame/support/procedural/src/default_no_bound.rs:89]: This is the intended syntax of `quote_spanned!`: https://docs.rs/quote/latest/quote/macro.quote_spanned.html#syntax

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2058182

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2058189

File: frame/support/procedural/src/default_no_bound.rs (hunk 2 above)
Comment (benluelo) [reply frame/support/procedural/src/default_no_bound.rs]: Turns out I had, just forgot to push those changes. That's what I get for writing code at 2am :sweat_smile:

File: frame/support/procedural/src/default_no_bound.rs (hunk 1 above)
Comment (benluelo) [reply frame/support/procedural/src/default_no_bound.rs]: See new implementation, I had to collect and re-iterate if `v.attrs` is borrowed.

File: frame/transaction-payment/asset-tx-payment/src/lib.rs [hunk 6]
```diff
@@ -97,6 +97,7 @@ pub(crate) type ChargeAssetLiquidityOf<T> =
 #[derive(Encode, Decode, DefaultNoBound, TypeInfo)]
 pub enum InitialPayment<T: Config> {
 	/// No initial fee was payed.
+	#[default]
```
Comment (bkchr) [inline frame/transaction-payment/asset-tx-payment/src/lib.rs:100]: Why did you change this?

File: frame/support/src/lib.rs [hunk 7]
```diff
@@ -685,7 +694,7 @@ pub use frame_support_procedural::crate_to_crate_version;
 #[macro_export]
 macro_rules! fail {
 	( $y:expr ) => {{
-		return Err($y.into())
+		return Err($y.into());
```
Comment (bkchr) [inline frame/support/src/lib.rs:697]:
```suggestion
		return Err($y.into())
```

File: frame/support/procedural/src/default_no_bound.rs [hunk 8]
```diff
@@ -15,82 +15,131 @@
...
-				match &first_variant.fields {
-					syn::Fields::Named(named) => {
-						let fields = named.named.iter().map(|i| &i.ident).map(|i| {
-							quote::quote_spanned!(i.span() =>
-								#i: core::default::Default::default()
-							)
-						});
+		Data::Enum(enum_) => {
+			if enum_.variants.is_empty() {
+				quote!(unreachable!())
```
Comment (bkchr) [inline frame/support/procedural/src/default_no_bound.rs:56]: This is not correct. You can not derive `Default` for an empty enum

File: frame/support/procedural/src/default_no_bound.rs [hunk 9]
```diff
@@ -15,82 +15,131 @@
...
+					.into_iter()
+					.flat_map(|variant| {
+						variant
+							.attrs
+							.iter()
+							.filter_map::<(), _>(|attr| attr.path.is_ident("default").then_some(()))
+							.collect::<Vec<_>>()
+							.into_iter()
+							.map(move |_| variant.clone())
+					})
```
Comment (bkchr) [inline frame/support/procedural/src/default_no_bound.rs:69]:
```suggestion
					.filter(|variant| {
						variant
							.attrs
							.iter()
							.any(|attr| attr.path.is_ident("default")
					})
					.cloned()
```

**Review by bkchr — COMMENTED** [review]
There are still some issues, but looking good otherwise.

File: frame/support/procedural/src/default_no_bound.rs (hunk 8 above)
Comment (benluelo) [reply frame/support/procedural/src/default_no_bound.rs]: For some reason I thought that was possible with std's derive, updated to return an error instead :+1:

File: frame/support/procedural/src/default_no_bound.rs (hunk 9 above)
Comment (benluelo) [reply frame/support/procedural/src/default_no_bound.rs]: This was grabbing all of the `#[default]` attributes with their respective variants (possibly duplicating a variant if it had multiple attributes). I changed the implementation because this was a bit confusing.

File: frame/transaction-payment/asset-tx-payment/src/lib.rs (hunk 6 above)
Comment (benluelo) [reply frame/transaction-payment/asset-tx-payment/src/lib.rs:100]: the `#[default]` attribute is now required:
```
error: no default declared, make a variant default by placing `#[default]` above it
  --> frame/transaction-payment/asset-tx-payment/src/lib.rs:98:10
   |
98 | pub enum InitialPayment<T: Config> {
   |          ^^^^^^^^^^^^^^
```

Comment (bkchr) [reply frame/transaction-payment/asset-tx-payment/src/lib.rs:100]: Ahh it was just deriving default using the first variant...

**Review by bkchr — APPROVED** [review]
Thank you!

File: frame/support/procedural/src/default_no_bound.rs [hunk 10]
```diff
@@ -15,82 +15,151 @@
...
-							)
-						});
-
-						quote::quote!( #name :: #ty_generics :: #variant_ident { #( #fields, )* } )
+		Data::Enum(enum_) => {
+			if enum_.variants.is_empty() {
+				return syn::Error::new_spanned(name, "cannot derive Default for an empty enum")
+					.to_compile_error()
+					.into();
+			} else {
```
Comment (KiChjang) [inline frame/support/procedural/src/default_no_bound.rs:59]:
```suggestion
			}
```

to reduce a level of indent.

**Review by KiChjang — APPROVED** [review]

**Review by sam0x17 — APPROVED** [review]

**Comment by bkchr** [comment]
bot rebase

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2083120
